python/lsst/multiprofit/modeller.py

```diff
@@ -0,0 +1,840 @@
+                if n_g != 1:
+                    raise ValueError(f"{component=} has {gaussians=} of len {n_g=}!=1")
+            param_fluxes = component.parameters(paramfilter=g2f.ParamFilter(nonlinear=False, channel=channel))
+            if len(param_fluxes) != 1:
+                raise ValueError(f"Can't make linear source from {component=} with {len(param_fluxes)=}")
+            param_flux: g2f.ParameterD = param_fluxes[0]
+            if param_flux.fixed:
+                gaussians_fixed.append(gaussians.at(0))
+            else:
+                gaussians_free.append((gaussians, param_flux))
+
+        return LinearGaussians(
+            gaussians_fixed=g2.Gaussians(gaussians_fixed), gaussians_free=tuple(gaussians_free)
+        )
+
+
+class FitInputsBase(ABC):
+    """Interface for inputs to a model fit."""
+
+    @abstractmethod
+    def validate_for_model(self, model: Model) -> list[str]:
+        """Check that this FitInputs is valid for a Model.
+
+        Parameters
+        ----------
+        model
+            The model to validate with.
+
+        Returns
+        -------
+        errors
+            A list of validation errors, if any.
+        """
+
+
+class FitInputsDummy(FitInputsBase):
+    """A dummy FitInputs that always fails to validate.
+
+    This class can be used to initialize a FitInputsBase that may be
+    reassigned to a non-dummy derived instance in a loop.
+    """
+
+    def validate_for_model(self, model: Model) -> list[str]:
+        return [
+            "This is a dummy FitInputs and will never validate",
+        ]
+
+
+class FitInputs(FitInputsBase, pydantic.BaseModel):
+    """Model fit inputs for gauss2dfit."""
+
+    model_config = arbitrary_allowed_config
+
+    jacobian: np.ndarray = pydantic.Field(None, title="The full Jacobian array")
+    jacobians: list[list[g2.ImageD]] = pydantic.Field(
+        title="Jacobian arrays (views) for each observation",
+    )
+    outputs_prior: tuple[g2.ImageD, ...] = pydantic.Field(
+        title="Jacobian arrays (views) for each free parameter's prior",
+    )
+    residual: np.ndarray = pydantic.Field(title="The full residual (chi) array")
+    residuals: list[g2.ImageD] = pydantic.Field(
+        default_factory=list,
+        title="Residual (chi) arrays (views) for each observation",
+    )
+    residuals_prior: g2.ImageD = pydantic.Field(
+        title="Shared residual array for all Prior instances",
+    )
+
+    @classmethod
+    def get_sizes(
+        cls,
+        model: Model,
+    ):
+        """Initialize Jacobian and residual arrays for a model.
+
+        Parameters
+        ----------
+        model : `gauss2d.fit.Model`
+            The model to initialize arrays for.
+        """
+        priors = model.priors
+        n_prior_residuals = sum(len(p) for p in priors)
+        params_free = tuple(get_params_uniq(model, fixed=False))
+        n_params_free = len(params_free)
+        # There's one extra validation array
```

Comment:
This is a "feature" of sorts in gauss2d_fit. The zeroth index is reserved to dump e.g. Jacobian terms for unneeded parameters. If you fix the centroid parameters, for example, it will still compute the Jacobian for it, since I did not implement any way to avoid doing that (there are already too many template parameters in the evaluator and a conditional could slow down the most common case of the majority of the parameters being free).

The Jacobian output is set up by specifying an index to output the value and a weight (which is how parameter transforms are accounted for). For a fixed parameter, you'd set the weight and the index to zero. Technically it's not necessary to set the index to zero as long as the weight is zero, but this makes it easier to validate that the weights were set correctly since you can check that all of the values in the zero-index image are exactly zero.

Reply:
Ok, maybe just explain some of that in the comment.